### ISSUE TITLE
Clarify purpose of slice.name in @reduxjs/toolkit docs

### ISSUE THREAD

**Issue by F1LT3R** [reduxjs/redux#3821]
CC: @markerikson

Firstly, thanks for Redux. I love it. I wanted to share a little of my learning experience and make a suggestion that I think will help developers from getting stuck when first using Redux with React.

I put off learning Redux for a long time, due to the lack of state-complexity of the apps I was building. I started digging into Vanilla Redux recently, pairing it with Vanilla Web Components (ie: no React).

When it came to applying Redux to React, things became much harder to reason about. This is surprising as the React Eco-System is generally super clear to understand.

What really got me stuck with @reduxjs/toolkit, was the reference to the slice. By this, I am referring to the name of the slice itself. I'm going to use the CodeSandbox example you guys provided to try and make this as clear as I can. https://codesandbox.io/s/github/reduxjs/redux-essentials-example-app/tree/checkpoint-3-postRequests/?from-embed

I would have assumed that the `name: "posts"` would be what controls this reference, but it does not! 😕

```js
// https://github.com/reduxjs/redux-essentials-example-app/blob/checkpoint-3-postRequests/src/features/posts/postsSlice.js#L23

const postsSlice = createSlice({
  name: 'posts',  // <- Renaming this has zero effect! 
  initialState,
```

When trying to remove the name, Redux Toolkit complains that "`name` is a required option for createSlice".

![CleanShot 2020-07-21 at 14 14 01@2x](https://user-images.githubusercontent.com/1218446/88091054-852d7500-cb5c-11ea-9539-dbf9c29a87fc.png)

> ### **🤔 Huh?**
> You _must_ declare `slice.name`, but `slice.name` has zero bearing on how you use Redux. (As far as I can tell).

The actual reference is found in `src/app/store.js`:

```js
// https://github.com/reduxjs/redux-essentials-example-app/blob/checkpoint-3-postRequests/src/app/store.js#L8

import { configureStore } from '@reduxjs/toolkit'

import postsReducer from '../features/posts/postsSlice'
import usersReducer from '../features/users/usersSlice'

export default configureStore({
  reducer: {
    posts: postsReducer,  // <-- THIS
    users: usersReducer,
  },
})
```

You can also see this confusing `posts.posts` reference here:

```js
//  https://github.com/reduxjs/redux-essentials-example-app/blob/checkpoint-3-postRequests/src/features/posts/postsSlice.js#L66

export const selectAllPosts = state => state.posts.posts;
```

The confusion continued for me when looking at the React component's reference to `state.posts` :

```jsx
// https://github.com/reduxjs/redux-essentials-example-app/blob/checkpoint-3-postRequests/src/features/posts/PostsList.js#L32

export const PostsList = () => {
  const dispatch = useDispatch()
  const posts = useSelector(selectAllPosts)

  // Does `state.posts` refer to the `posts[]` array of the slice's state as I expected?
  // Nope! It refers to the post's reference on the store.
  const postStatus = useSelector((state) => state.posts.status)
  const error = useSelector((state) => state.posts.error)

  useEffect(() => {
    if (postStatus === 'idle') {
      dispatch(fetchPosts())
    }
  }, [postStatus, dispatch])
```

This made things very difficult to reason about in the context of React, where as Vanilla Redux was as clear as day.

The React community prides itself on declarative code for good reason. It falls down here, for me. Perhaps I missed something somewhere, but I don't see what the point of `slice.name` is. This makes it much harder to figure out what is going on when hooking up Redux to React.

To recap, the user is required to create three "things" called "posts", yet only two of them seem to be of value to the end user.

1. `export default configureStore({reducer: { posts:... }})`
2. `const postsSlice = createSlice({name: 'posts'})`
3. `const initialState = {posts: []}`

- Is it possible to infer the `slice.name` automatically from the store?
- Can you clarify what is the purpose of `slice.name` for the end user?

**Comment by markerikson** [comment]
The `name` field of `createSlice` specifically is used as a prefix for all action types generated within this slice:

```js
const slice = createSlice({
  name: "todos",
  initialState: [],
  reducers: {
    todoAdded(state, action) {}
  }
})

console.log(slice.actions.todoAdded("Buy milk"));
// {type: "todos/todoAdded", payload: "Buy milk"}
```

You're correct that it has no actual direct relation to the name of the field in the root state tree, because that is entirely defined by what your root reducer returns.  That root reducer is normally generated by `combineReducers`.  So, for example, I could take my `todosReducer`, and put it at _any_ field name in the root state:

```js
const rootReducer = combineReducers({
  someOtherFieldEntirely: todosReducer
})

console.log(rootReducer(undefined, {type: "initAction"}));
// {someOtherFieldEntirely: []}
```

However, it is _likely_ that there will be a correlation between the `slice.name` field and the name of the root state field for this reducer, because people like naming things the same way:

```js
const rootReducer = combineReducers({
  todos: todosReducer
})
```

It's _not_ possible to infer the "name", for multiple reasons:

- The name has to be known at the time the slice is defined, because it's used internally when generating the action creators
- Per above, there is no guarantee that a slice named "x" will be mounted as `state.x`
- It's even possible that a given slice reducer may be mounted as a _nested_ reducer in the state tree, like:

```js
const modalsSlice = createSlice({
  name: "modals",
  initialState,
  reducers
})

const uiReducer = combineReducers({
  modals: modalsReducer,
  dialogs: dialogsReducer
})

const rootReducer = combineReducers({
  users: usersReducer,
  ui: uiReducer
})
```

Note that the tutorial sequence _does_ tell you the purpose of `slice.name`, although it's not bolded or specifically highlighted:

https://redux.js.org/tutorials/essentials/part-2-app-structure#creating-slice-reducers-and-actions

> Redux Toolkit has a function called `createSlice`, which takes care of the work of generating action type strings, action creator functions, and action objects. All you have to do is define a `name` for this slice, write an object that has some reducer functions in it, and it generates the corresponding action code automatically. The string from the name option is used as the first part of each action type, and the key name of each reducer function is used as the second part. So, the `"counter"` name + the `"increment"` reducer function generated an action type of `{type: "counter/increment"}`. (After all, why write this by hand if the computer can do it for us!)


The `state.posts.posts` piece you're pointing to is an entirely separate point of concern.

Within a given slice, you can have whatever further nested state you want.  In this case, the `postsSlice` started with _only_ storing an array of post entries.  Later on, we started tracking loading state for the slice as well as we fetch posts from the server.

At that point, we now need to track an array of post objects, _and_ loading state, so we can't _just_ have an array.  So, the state for the `postsSlice` switching to being an object that has an array and some other loading state fields in it.

We can name the nested array anything we want, but for consistency, we can still call it `posts`.  Since the entire `postsReducer` is being used to handle all updates to the root `state.posts`, we now end up with the slightly awkward naming situation of `state.posts.posts`, because the nested array is inside the slice state.

That naming quirk gets resolved a bit later in the tutorial when we switch over to using the `createEntityAdapter` API, which stores data as `{ids: [], entities: {}}`.  So, you've got `state.posts.entities` from there on out.

You might want to read through the [Initializing State](https://redux.js.org/recipes/structuring-reducers/initializing-state) docs page, which recaps how some of this process works, as well as the rest of the [Structuring Reducers](https://redux.js.org/recipes/structuring-reducers/structuring-reducers) section.

I'm a bit confused why you're saying this confusion is React-related.  None of these concepts have to do with React.  `useSelector` receives the entire Redux root state tree as its argument.  So, depending on the progress through the tutorial, you read `state => state.posts`, `state => state.posts.posts`, etc, as the structure of the state tree changes.


Hopefully that clarifies the behavior.  If you've got any further questions, or suggestions for how to specifically improve these explanations in the tutorial or docs, please let me know.

**Comment by F1LT3R** [comment]
I appreciate the clarification, that's really helpful - thanks @markerikson!

> I'm a bit confused why you're saying this confusion is React-related. None of these concepts have to do with React. 

I'm unsure why you're saying this. Perhaps you are saying that this problem would be exactly the same if I used @reduxjs/toolkit without React? (I didn't think the toolkit worked without React).

**Comment by markerikson** [comment]
Exactly.  Both the Redux core, and Redux Toolkit, are 100% UI agnostic plain JS libraries.  They can be used with _any_ UI layer, it's just that React is the most common.

Everything about the "store", "reducers", "actions", and "slices", is unrelated to the UI.

So, everything that I said about `state.posts.posts`, etc, is true regardless of what UI layer you're using.

**Comment by F1LT3R** [comment]
Awesome, thanks @markerikson
